File: src/main/java/bg/softuni/bookshelf/service/bookshelf/BookshelfService.java
package bg.softuni.bookshelf.service.bookshelf;

import bg.softuni.bookshelf.service.book.dto.BookSummaryDto;
import bg.softuni.bookshelf.service.bookshelf.dto.*;
import bg.softuni.bookshelf.shared.dto.PagedResponse;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

/**
 * Service interface for managing a user's personal bookshelves.
 * <p>
 * This contract defines all business operations related to creating, managing,
 * and viewing custom bookshelves and their contents.
 */
public interface BookshelfService {

    /**
     * Creates a new, empty bookshelf for a specific user.
     * <p>
     * This method follows the "write-through-get" pattern. It returns the full
     * state of the newly created resource to prevent the client from needing to
     * make a subsequent API call, which improves performance and simplifies client-side logic.
     *
     * @param createDto The DTO containing the new shelf's details.
     * @param ownerId   The UUID of the user who will own this shelf.
     * @return A {@link BookshelfDetailsDto} representing the newly created bookshelf.
     */
    BookshelfDetailsDto createShelf(BookshelfCreateDto createDto, UUID ownerId);

    /**
     * Retrieves a paginated list of all bookshelves owned by a specific user.
     *
     * @param userId   The UUID of the user.
     * @param pageable The pagination and sorting information.
     * @return A {@link PagedResponse} of {@link BookshelfSummaryDto} objects.
     */
    PagedResponse<BookshelfSummaryDto> getShelvesForUser(UUID userId, Pageable pageable);

    /**
     * Retrieves the detailed view of a single bookshelf, including a paginated list of its books.
     *
     * @param shelfId The UUID of the bookshelf.
     * @return A {@link BookshelfDetailsDto} containing the shelf's details and its books.
     */
    BookshelfDetailsDto getShelfById(UUID shelfId);

    /**
     * Retrieves a paginated list of all books contained within a specific bookshelf.
     *
     * @param shelfId  The UUID of the bookshelf.
     * @param pageable The pagination and sorting information for the books.
     * @return A {@link PagedResponse} of {@link BookSummaryDto} objects.
     */
    PagedResponse<BookSummaryDto> getBooksInShelf(UUID shelfId, Pageable pageable);

    /**
     * Updates the details (e.g., name, description) of an existing bookshelf.
     *
     * @param shelfId   The UUID of the bookshelf to update.
     * @param updateDto The DTO containing the new details.
     * @return A {@link BookshelfDetailsDto} representing the updated state of the shelf.
     */
    BookshelfDetailsDto updateShelf(UUID shelfId, BookshelfUpdateDto updateDto);

    /**
     * Deletes a bookshelf. This action is permanent and will also remove all
     * associations between this shelf and any books it contained.
     *
     * @param shelfId The UUID of the bookshelf to delete.
     */
    void deleteShelf(UUID shelfId);

    /**
     * Adds a book to a specific bookshelf.
     *
     * @param shelfId    The UUID of the bookshelf.
     * @param addBookDto The DTO containing the UUID of the book to add.
     */
    void addBookToShelf(UUID shelfId, AddBookToBookshelfDto addBookDto);

    /**
     * Removes a book from a specific bookshelf.
     *
     * @param shelfId The UUID of the bookshelf.
     * @param bookId  The UUID of the book to remove.
     */
    void removeBookFromShelf(UUID shelfId, UUID bookId);
}

File: frontend/src/app/core/services/bookshelf.service.ts
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import {
  AddBookToBookshelfDto,
  BookshelfCreateDto,
  BookshelfDetailsDto,
  BookshelfUpdateDto,
  Pageable,
  PagedResponseBookshelfSummaryDto,
  PagedResponseBookSummaryDto,
  UserShelfAPIService,
} from '../../api';

/**
 * Service responsible for managing user bookshelves.
 * Acts as a facade over the generated OpenAPI client (`UserShelfAPIService`).
 */
@Injectable({ providedIn: 'root' })
export class BookshelfService {
  private readonly userShelfApiService = inject(UserShelfAPIService);

  /**
   * Retrieves a paginated list of bookshelves belonging to the currently authenticated user.
   * @param pageable Pagination and sorting configuration.
   * @returns An observable emitting a paginated result of bookshelf summaries.
   */
  getShelvesForUser(pageable: Pageable): Observable<PagedResponseBookshelfSummaryDto> {
    return this.userShelfApiService.getUserShelves(pageable);
  }

  /**
   * Creates a new bookshelf for the currently authenticated user.
   * @param createDto Data required to create the bookshelf.
   * @returns An observable emitting the detailed view of the newly created bookshelf.
   */
  createShelf(createDto: BookshelfCreateDto): Observable<BookshelfDetailsDto> {
    return this.userShelfApiService.createShelf(createDto);
  }

  /**
   * Retrieves the detailed view of a specific bookshelf.
   * @param shelfId The UUID of the bookshelf to retrieve.
   * @returns An observable emitting the detailed bookshelf data.
   */
  getShelfById(shelfId: string): Observable<BookshelfDetailsDto> {
    return this.userShelfApiService.getShelfById(shelfId);
  }

  /**
   * Retrieves a paginated list of books contained within a specific bookshelf.
   * @param shelfId The UUID of the bookshelf.
   * @param pageable Pagination and sorting configuration.
   * @returns An observable emitting a paginated result of book summaries.
   */
  getBooksInShelf(shelfId: string, pageable: Pageable): Observable<PagedResponseBookSummaryDto> {
    return this.userShelfApiService.getBooksInShelf(shelfId, pageable);
  }

  /**
   * Updates an existing bookshelf.
   * @param shelfId The UUID of the bookshelf to update.
   * @param updateDto Data required to update the bookshelf.
   * @returns An observable emitting the updated detailed view of the bookshelf.
   */
  updateShelf(shelfId: string, updateDto: BookshelfUpdateDto): Observable<BookshelfDetailsDto> {
    return this.userShelfApiService.updateShelf(shelfId, updateDto);
  }

  /**
   * Deletes a specific bookshelf.
   * @param shelfId The UUID of the bookshelf to delete.
   * @returns An observable that completes when the deletion is successful.
   */
  deleteShelf(shelfId: string): Observable<void> {
    return this.userShelfApiService.deleteShelf(shelfId);
  }

  /**
   * Adds a book to a specific bookshelf.
   * @param shelfId The UUID of the bookshelf.
   * @param addBookDto Data containing the ID of the book to add.
   * @returns An observable that completes when the book is added successfully.
   */
  addBookToShelf(shelfId: string, addBookDto: AddBookToBookshelfDto): Observable<void> {
    return this.userShelfApiService.addBookToShelf(shelfId, addBookDto);
  }

  /**
   * Removes a book from a specific bookshelf.
   * @param shelfId The UUID of the bookshelf.
   * @param bookId The UUID of the book to remove.
   * @returns An observable that completes when the book is removed successfully.
   */
  removeBookFromShelf(shelfId: string, bookId: string): Observable<void> {
    return this.userShelfApiService.removeBookFromShelf(shelfId, bookId);
  }
}
